sdks/rust/src/sdk.rs

```diff
@@ -12,349 +12,211 @@
-                Err(e) => {
-                    if counter > 30 {
-                        return Err(ErrorKind::Grpc(e).into());
-                    }
-                    sleep(Duration::from_secs(1));
-                    continue;
-                }
+            connect_interval.tick().await;
+
+            if client.get_game_server(empty()).await.is_ok() {
+                break;
             }
         }
 
-        // Keep both sender and receiver as RPC is canceled when sender or receiver is dropped
-        let (sender, receiver) = cli.health()?;
-        Ok(Sdk {
-            client: Arc::new(cli),
-            health: Arc::new(Mutex::new(Some(sender))),
-            health_receiver: Arc::new(Mutex::new(Some(receiver))),
-            alpha: Arc::new(alpha),
+        Ok(Self {
+            client,
+            alpha,
         })
     }
 
     /// Alpha returns the Alpha SDK
-    pub fn alpha(&self) -> &Arc<Alpha> {
+    #[inline]
+    pub fn alpha(&self) -> &Alpha {
         &self.alpha
     }
 
     /// Marks the Game Server as ready to receive connections
-    pub fn ready(&self) -> Result<()> {
-        let req = sdk::Empty::default_instance();
-        let res = self.client.ready(req).map(|_| ())?;
-        Ok(res)
+    #[inline]
+    pub async fn ready(&mut self) -> Result<()> {
+        Ok(self.client.ready(empty()).await.map(|_| ())?)
     }
 
     /// Allocate the Game Server
-    pub fn allocate(&self) -> Result<()> {
-        let req = sdk::Empty::default_instance();
-        let res = self.client.allocate(req).map(|_| ())?;
-        Ok(res)
+    #[inline]
+    pub async fn allocate(&mut self) -> Result<()> {
+        Ok(self.client.allocate(empty()).await.map(|_| ())?)
     }
 
     /// Marks the Game Server as ready to shutdown
-    pub fn shutdown(&self) -> Result<()> {
-        let req = sdk::Empty::default_instance();
-        let res = self.client.shutdown(req).map(|_| ())?;
-        Ok(res)
-    }
+    #[inline]
+    pub async fn shutdown(&mut self) -> Result<()> {
+        Ok(self.client.shutdown(empty()).await.map(|_| ())?)
+    }
+
+    /// Creates a task that sends a health ping to the SDK server on every interval
+    /// tick. It is recommended to only have 1 of these at a time.
+    #[inline]
+    pub fn spawn_health_task(&self, interval: Duration) -> tokio::sync::oneshot::Sender<()> {
+        let mut health_client = self.clone();
+        let (tx, mut rx) = tokio::sync::oneshot::channel();
+
+        tokio::task::spawn(async move {
+            let health_stream = async_stream::stream! {
+                let mut health_interval = tokio::time::interval(interval);
+                loop {
+                    tokio::select! {
+                        _ = health_interval.tick() => {
+                            yield empty();
+                        }
+                        _ = &mut rx => {
+                            break;
+                        }
+                    }
+                }
+            };
```

Comment:
The previous health check code was kind of strange IMO, since this client internally created a stream, but then required the caller to actually still invoke the `health` method themselves to send the health checks, which seemed error prone. I opted for a slightly more battery included approach of just having a caller specify an interval they want health checks to be sent at, and return a `oneshot` that they can use to cancel/drop the health check task.

The one unfortunate thing with this change is that we lose the error status on the `client.health()` call, but I figured this was ok.

Reply:
Unfortunately, this doesn't actually test health-iness, because the implementer can't attach any game server specific health checking to the code. This only checks to make sure the SDK is remaining healthy.

I know we can send in a stop request - but then then the end user still needs to implement a regular health checking loop anyway - so I would posit that it's easier to have a simple `health()` request, so it can be integrated as needed with each backing code. 

Thoughts?

Reply:
Yah, this was my bad, I was only thinking of API niceness not actual functionality with this. I changed it to instead be a bounded async channel, so the user just needs to do

```rust
let health_check = sdk.health_check();

loop {
    // If async
    health_check.send(()).await.expect("health check dropped");
    // If non-async
    if let Err(tokio::sync::mpsc::error::TrySendError::Closed(_)) = health_check.try_send(()) {
        panic!("health check receiver was closed");
    }
}
```